sql: broken tests in Bazel

We are making a lot of progress in the migration from make to Bazel. Before we can complete our migration, we need to make sure that all tests in-tree continue to pass when run inside of the Bazel sandbox. The following tests are broken when run in the Bazel sandbox:

    pkg/sql/opt/memo:memo_test
    pkg/sql/opt/norm:norm_test
    pkg/sql/opt/xform:xform_test

Please help us by doing the following:

* Reproduce the failure locally; for example, with `bazel test pkg/sql/opt/xform:xform_test`, and consult the test logs to see what went wrong
* Identify the root cause of the failure
* If you are able, repair the failure such that the test passes, and submit a PR removing the `broken_in_bazel` tag from the test (see the linked documentation below)
* If not, take your findings to #bazel in Slack, and we’ll work with you to figure out how we can get the tests working

Helpful documentation:
* [How to ensure your tests can run in the Bazel sandbox (consider starting here!)](https://cockroachlabs.atlassian.net/wiki/spaces/CRDB/pages/1553172016/How+to+ensure+your+tests+can+run+in+the+Bazel+sandbox)
* [How to ensure your code builds with Bazel](https://cockroachlabs.atlassian.net/wiki/spaces/CRDB/pages/1380090083/How+to+ensure+your+code+builds+with+Bazel)
* [A user's guide to Bazel](https://docs.bazel.build/versions/4.0.0/guide.html)

```
exec ${PAGER:-/usr/bin/less} "$0" || exit 1
Executing tests from //pkg/sql/opt/xform:xform_test
-----------------------------------------------------------------------------
I210312 19:29:22.191994 1 (gostd) rand.go:92  [-] 1  random seed: -539133880944181066
--- FAIL: TestExternal (0.18s)
    test_log_scope.go:73: test logs captured to: /tmp/logTestExternal363374037
    test_log_scope.go:74: use -show-logs to present logs inline
    --- FAIL: TestExternal/postgis-tutorial (0.00s)
        opt_tester.go:1331: open pkg/sql/opt/testutils/opttester/testfixtures/postgis_tutorial_schema: no such file or directory
    --- FAIL: TestExternal/postgis-tutorial-idx (0.00s)
        opt_tester.go:1331: open pkg/sql/opt/testutils/opttester/testfixtures/postgis_tutorial_schema_indexes: no such file or directory
    --- FAIL: TestExternal/tpcc (0.00s)
        opt_tester.go:1331: open pkg/sql/opt/testutils/opttester/testfixtures/tpcc_schema: no such file or directory
    --- FAIL: TestExternal/tpcc-later-stats (0.00s)
        opt_tester.go:1331: open pkg/sql/opt/testutils/opttester/testfixtures/tpcc_schema: no such file or directory
    --- FAIL: TestExternal/tpcc-no-stats (0.00s)
        opt_tester.go:1331: open pkg/sql/opt/testutils/opttester/testfixtures/tpcc_schema: no such file or directory
    --- FAIL: TestExternal/tpce (0.00s)
        opt_tester.go:1331: open pkg/sql/opt/testutils/opttester/testfixtures/tpce_schema: no such file or directory
    --- FAIL: TestExternal/tpce-no-stats (0.00s)
        opt_tester.go:1331: open pkg/sql/opt/testutils/opttester/testfixtures/tpce_schema: no such file or directory
    --- FAIL: TestExternal/tpch (0.00s)
        opt_tester.go:1331: open pkg/sql/opt/testutils/opttester/testfixtures/tpch_schema: no such file or directory
    --- FAIL: TestExternal/tpch-no-stats (0.00s)
        opt_tester.go:1331: open pkg/sql/opt/testutils/opttester/testfixtures/tpch_schema: no such file or directory
    --- FAIL: TestExternal/ycsb (0.00s)
        opt_tester.go:1331: open pkg/sql/opt/testutils/opttester/testfixtures/ycsb_schema: no such file or directory
    optimizer_test.go:219: -- test log scope end --
FAIL
```

Will follow the docs to add the missing deps.

--- a/pkg/sql/opt/memo/BUILD.bazel
+++ b/pkg/sql/opt/memo/BUILD.bazel
@@ -60,9 +60,9 @@ go_test(
     ],
     data = glob(["testdata/**"]) + [
         "@cockroach//c-deps:libgeos",
+        "//pkg/sql/opt/testutils/opttester:testfixtures",
     ],
     embed = [":memo"],
-    tags = ["broken_in_bazel"],
     deps = [
         "//pkg/settings/cluster",
         "//pkg/sql/inverted",

--- a/pkg/sql/opt/norm/BUILD.bazel
+++ b/pkg/sql/opt/norm/BUILD.bazel
@@ -66,9 +66,9 @@ go_test(
     ],
     data = glob(["testdata/**"]) + [
         "@cockroach//c-deps:libgeos",
+        "//pkg/sql/opt/testutils/opttester:testfixtures",
     ],
     embed = [":norm"],
-    tags = ["broken_in_bazel"],
     deps = [
         "//pkg/settings/cluster",
         "//pkg/sql/opt",

--- a/pkg/sql/opt/testutils/opttester/BUILD.bazel
+++ b/pkg/sql/opt/testutils/opttester/BUILD.bazel
@@ -14,6 +14,7 @@ go_library(
     importpath = "github.com/cockroachdb/cockroach/pkg/sql/opt/testutils/opttester",
     visibility = ["//visibility:public"],
     deps = [
+        "//pkg/build/bazel",
         "//pkg/roachpb",
         "//pkg/security",
         "//pkg/settings/cluster",
@@ -54,4 +55,10 @@ go_test(
         "//pkg/sql/opt/testutils/testcat",
         "@com_github_cockroachdb_datadriven//:datadriven",
     ],
+)
+
+filegroup(
+    name = "testfixtures",
+    srcs = glob(["testfixtures/**"]),
+    visibility = ["//pkg/sql/opt:__subpackages__"],
 )

--- a/pkg/sql/opt/testutils/opttester/opt_tester.go
+++ b/pkg/sql/opt/testutils/opttester/opt_tester.go
@@ -28,6 +28,7 @@ import (
 	"text/tabwriter"
 	"time"
 
+	"github.com/cockroachdb/cockroach/pkg/build/bazel"
 	"github.com/cockroachdb/cockroach/pkg/roachpb"
 	"github.com/cockroachdb/cockroach/pkg/security"
 	"github.com/cockroachdb/cockroach/pkg/settings/cluster"
@@ -1322,16 +1323,28 @@ func (ot *OptTester) Import(tb testing.TB) {
 	if ot.Flags.File == "" {
 		tb.Fatal("file not specified")
 	}
-	// Find the file to be imported in opttester/testfixtures.
-	_, optTesterFile, _, ok := runtime.Caller(1)
-	if !ok {
-		tb.Fatalf("unable to find file %s", ot.Flags.File)
-	}
-	path := filepath.Join(filepath.Dir(optTesterFile), "testfixtures", ot.Flags.File)
+	path := ot.testFixturePath(tb, ot.Flags.File)
 	datadriven.RunTest(tb.(*testing.T), path, func(t *testing.T, d *datadriven.TestData) string {
 		tester := New(ot.catalog, d.Input)
 		return tester.RunCommand(t, d)
 	})
+}
+
+// testFixturePath returns the path of a fixture inside opttester/testfixtures.
+func (ot *OptTester) testFixturePath(tb testing.TB, file string) string {
+	if bazel.BuiltWithBazel() {
+		runfile, err := bazel.Runfile("pkg/sql/opt/testutils/opttester/testfixtures/" + file)
+		if err != nil {
+			tb.Fatalf("%s; is your package missing a dependency on \"//pkg/sql/opt/testutils/opttester:testfixtures\"?", err)
+		}
+		return runfile
+	}
+	// Get the path to this file from the runtime.
+	_, thisFilePath, _, ok := runtime.Caller(0)
+	if !ok {
+		tb.Fatal("unable to get caller information")
+	}
+	return filepath.Join(filepath.Dir(thisFilePath), "testfixtures", file)
 }
 
 // InjectStats constructs and executes an ALTER TABLE INJECT STATISTICS

--- a/pkg/sql/opt/xform/BUILD.bazel
+++ b/pkg/sql/opt/xform/BUILD.bazel
@@ -62,9 +62,9 @@ go_test(
     ],
     data = glob(["testdata/**"]) + [
         "@cockroach//c-deps:libgeos",
+        "//pkg/sql/opt/testutils/opttester:testfixtures",
     ],
     embed = [":xform"],
-    tags = ["broken_in_bazel"],
     deps = [
         "//pkg/config/zonepb",
         "//pkg/roachpb",

